File: internal/api/gallery.go
package api

import (
	"net/http"
	"strings"

	"github.com/janishar/helmstudio/internal/api/studioapi"
)

// The launcher's gallery (03 §10, docs/decisions.md 2026-09-19).
//
// The three operations under /launcher/ are the studio-api gallery and asset
// reads served to the daemon's own page. Nothing about them is implemented
// here: the request is re-pointed at the studio-api path it mirrors and
// answered by the same generated router over the same service, under a
// principal that is the launcher rather than a studio. So the parameter
// decoding, Q9's read rule, the filter digest, the paging and the error shapes
// are the ones a studio gets, and there is no second copy of them to drift.
//
// What keeps that from being "the studio API with no token" is the mux: only
// the three patterns below reach this handler, and a path it does not have is
// the launcher's 404 like any other.

// routeGallery mounts them. Without a service there is no store to read.
func (s *Server) routeGallery() {
	if s.service == nil {
		return
	}
	h := studioapi.NewHandler(s.service, studioapi.Fixed(studioapi.Launcher()), s.logf)
	for _, pattern := range []string{
		"GET " + Base + "/launcher/gallery/items",
		"GET " + Base + "/launcher/assets/{id}",
		"GET " + Base + "/launcher/assets/{id}/thumb",
	} {
		s.mux.Handle(pattern, asStudioAPI(h))
	}
}

// asStudioAPI drops the /launcher segment and asks the studio-api router for
// the operation underneath. The gallery query is forced to every studio: the
// launcher has no items of its own, so `scope=self` would answer an empty page
// rather than the screen anybody asked for, and the contract gives it no scope
// to send.
func asStudioAPI(h *studioapi.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		u := *r.URL
		u.Path = strings.Replace(u.Path, Base+"/launcher/", Base+"/", 1)
		if u.RawPath != "" {
			u.RawPath = strings.Replace(u.RawPath, Base+"/launcher/", Base+"/", 1)
		}
		if strings.HasSuffix(u.Path, "/gallery/items") {
			q := u.Query()
			q.Set("scope", "all")
			u.RawQuery = q.Encode()
		}
		r2 := r.Clone(r.Context())
		r2.URL = &u
		h.ServeHTTP(w, r2)
	})
}

File: web/gallery.js
// The launcher's Gallery (03 §10, amended 2026-09-19): every studio's work in
// one grid, and the same grid scoped to one studio beside that studio's page.
//
// The grid is <helm-gallery> — the component a studio mounts in its own page —
// given the launcher's client, which reads every studio because the daemon
// answers its three operations under a principal that holds gallery.read_all.
// One query serves both views, which is what 03 §10 asks of them.
//
// What helmstudio draws around it is the one filter a component cannot have:
// which studio made it. A component is given a client and never learns what is
// installed (04 §11 rule 2), so the chips are the launcher's and the choice
// reaches the component as its `studio` attribute.
//
// The launcher's client has no event stream and no gallery update, so there is
// no live insertion and no star here. Both are 04 §9's smaller component
// rather than a broken one: an item belongs to the studio that made it, and a
// PATCH of another's is refused, so a star the launcher drew could only fail.

import { el } from "./ui.js";

// What a take is called once it leaves helmstudio. The mime is on the item, so
// naming the file costs no request; anything unmapped keeps the mime's own
// subtype, which is right often enough and never empty.
const EXTENSION = {
  "image/png": "png", "image/jpeg": "jpg", "image/webp": "webp", "image/gif": "gif",
  "video/mp4": "mp4", "video/quicktime": "mov", "video/webm": "webm",
  "audio/wav": "wav", "audio/x-wav": "wav", "audio/mpeg": "mp3", "audio/flac": "flac", "audio/ogg": "ogg",
};

function saveName(item) {
  const asset = item.asset || {};
  const mime = asset.mime || "";
  const ext = EXTENSION[mime] || mime.split("/")[1] || "bin";
  const stem = String(item.title || item.id).replace(/[^\w.-]+/g, "-").replace(/^-+|-+$/g, "").slice(0, 60);
  return (stem || "take") + "." + ext;
}

/**
 * grid keeps one <helm-gallery> per key across redraws. A poll runs every two
 * seconds; a gallery rebuilt on each one would lose its scroll, its filters
 * and whatever is playing in its viewer.
 *
 * Save is the launcher's own, in the viewer's `viewer-actions` slot — beside
 * Full screen, on the item being looked at, which is where someone decides
 * they want to keep it. What can be done with an item is not a component's
 * business (04 §11 rule 5). It
 * is a link with `download` rather than a fetch, so the browser and the Mac
 * app take the same path — WebKit turns it into a `WKDownload` the app's
 * delegate writes to ~/Downloads, and the context menu's own Download Image,
 * which reaches neither, stops being the only way to keep a render.
 */
export function grid(ctx, key, studioID) {
  const gallery = ctx.keep("gallery:" + key, () => {
    const g = document.createElement("helm-gallery");
    // Every studio, always: the launcher has no items of its own, and its
    // operation has no scope to send anyway.
    g.setAttribute("scope", "all");
    g.client = ctx.client;
    const save = el("a", { class: "helm-btn helm-gallery-save", slot: "viewer-actions", hidden: true, download: "" });
    save.textContent = "Save";
    g.append(save);
    g.addEventListener("select", (e) => {
      const item = (e.detail && e.detail.item) || null;
      if (!item || !item.asset_id) {
        save.hidden = true;
        return;
      }
      save.href = "/api/v1/launcher/assets/" + encodeURIComponent(item.asset_id);
      save.download = saveName(item);
      save.hidden = false;
    });
    return g;
  });
  if (studioID) gallery.setAttribute("studio", studioID);
  else gallery.removeAttribute("studio");
  return gallery;
}

/**
 * chips are the scope: All studios, then one per installed studio in its own
 * hue. They are links, so a scope is a place you can come back to.
 */
export function chips(ctx, current, href) {
  // Anything that has ever been installed, which is not the "ready or
  // update_available" test the action buttons use: a studio whose build broke
  // this morning still made everything it made before that, and a chip is a
  // filter over what exists rather than a statement about what can run.
  const studios = (ctx.store.studios || []).filter((s) => s.install_state && s.install_state !== "listed");
  const one = (id, label, hue) => el("a", {
    class: "helm-chip helm-scope" + (id === current ? " helm-scope-on" : ""),
    href: href(id),
    "aria-current": id === current ? "true" : undefined,
    style: hue && hue.dark && hue.light ? `--_hue-dark: ${hue.dark}; --_hue-light: ${hue.light}` : undefined,
  }, hue ? el("span", { class: "helm-hue-dot" }) : null, label);
  return [
    one("", "All studios", null),
    ...studios.map((s) => one(s.id, s.name, s.hue)),
  ];
}

/** galleryScreen is the nav's Gallery: every studio, narrowed by a chip. */
export function galleryScreen(ctx) {
  const studioID = ctx.query.get("studio") || "";
  return el("div", { class: "helm-stack helm-gallery-screen" },
    el("div", { class: "helm-page-header" },
      el("h1", { class: "helm-title", text: "Gallery" }),
      ...chips(ctx, studioID, (id) => (id ? `#/gallery?studio=${encodeURIComponent(id)}` : "#/gallery"))),
    grid(ctx, "launcher", studioID));
}

/**
 * pickAsset opens the gallery as a picker over every studio and resolves to
 * the asset chosen, or null. What goes into a sequence is the page's choice,
 * not the editor's (04 §11 rule 5): `helm-timeline` asks, and this answers.
 */
export function pickAsset(ctx) {
  return new Promise((resolve) => {
    let chosen = null;
    const gallery = document.createElement("helm-gallery");
    gallery.setAttribute("scope", "all");
    gallery.setAttribute("picker", "");
    gallery.client = ctx.client;
    gallery.addEventListener("pick", (e) => {
      chosen = e.detail && e.detail.asset;
      node.close();
    });
    const node = el("dialog", { class: "helm-dialog helm-picker", "aria-labelledby": "pick-title" },
      el("h2", { class: "helm-dialog-title", id: "pick-title", text: "Add from the gallery" }),
      gallery,
      el("div", { class: "helm-dialog-actions" },
        el("button", { class: "helm-btn", type: "button", text: "Cancel", onclick: () => node.close() })));
    node.addEventListener("close", () => {
      node.remove();
      resolve(chosen);
    });
    document.body.append(node);
    node.showModal();
  });
}
